### ISSUE TITLE
Create Frame from generator

### ISSUE THREAD

**Issue by amachanic** [h2oai/datatable#3249]
Possible to allow the creation of a Frame from a generator? A lot of APIs will output a generator and it seems strange to have to materialize that data into a list just to then re-materialize that list into a Frame.

**Comment by samukweku** [comment]
@amachanic creating a Frame would require materialisation from the generator either ways, right?

**Comment by amachanic** [comment]
Yes, agreed with your final point. At the end of the day this is all about memory -- it's always at a premium in Python, which is why we (on my end at least) often exploit generators. But it does sound like there are some Frame internals that make this a bigger lift than I expected; thanks for sharing that insight.

**Comment by oleksiyskononenko** [comment]
You're welcome. I think this is doable, but may have some technical difficulties. I don't have experience with generators in pandas, what happens if the generator passed is infinite?

**Comment by amachanic** [comment]
It did pretty much what I expected: It sat there for a long time until it used up all of the RAM, and then the OS had enough and took it down.

![image](https://user-images.githubusercontent.com/39809964/159560693-551052ac-90ec-4303-aff5-bc335029b593.png)

**Comment by oleksiyskononenko** [comment]
@amachanic I've measured memory consumption of these two code blocks

```python
import pandas as pd
PD = pd.DataFrame([i for i in range(10**8)])
```
and
```python
import pandas as pd
PD = pd.DataFrame((i for i in range(10**8)))
```
and found that it is almost identical. It means that even though pandas allows frame creation from generators, it materializes generator to something like a list behind the scene. So there is no gain in memory when you use generator as a source. Could you please double check it on your side? 

My feeling is that this could be implemented in a one-by-one fashion: requesting one element from a generator, converting it to C++ primitive, do it again — but it may not be efficient from the performance point of view.
